walkthroughs/1C-managing-and-protecting-using-api-mgmt/walkthrough.adoc

```diff
@@ -0,0 +1,381 @@
+
+=== Configure an Application Plan and an Application
+
+. Create a new *Application Plan*:
+.. Select *Applications > Application Plans* from the side navigation.
+.. Click *Create Application Plan*.
+.. Enter `orders-api-plan` for *Name* and *System name*.
+.. Leave the other fields with their default values.
+.. Click *Create Application Plan*. You will be redirected to the *Application Plans* screen.
+.. Select the *Publish* button, beside your plan list item, to publish the Plan.
++
+. Select the `orders-api-plan` plan name in the list to return to the edit screen.
++
+. Set a limit of 5 calls per hour:
+.. From the *Metrics, Methods, Limits & Pricing Rules* section, click the *Limits (0)* button.
+.. Click the *New usage limit* button.
+.. Set the *Period* to *hour*.
+.. Set the *Max. value* to *5*.
+.. Click *Create usage limit*.
++
+. Create a new *Application* for the *Developer* account, assigned to the Plan:
+.. Select *Audience* from the top navigation bar dropdown.
+.. Select the *Developer* account to open the *Account Summary* page.
+.. Select the *(num) Application* item from the breadcrumbs to view Applications.
+.. Click the *Create Application* button in the top right.
+.. Select the `orders-api-plan` Plan in the *Application plan* dropdown.
+.. Enter `orders-api-app` in the *Name* and *Description* fields.
+.. Click *Create Application*. You will be redirected to the application details page.
+.. Note the *Client ID* and *Client Secret* that are generated automatically.
+When the application is created, 3scale should create a client for that application in {sso-ProductName} realm.
++
+. Verify that a client has been created in {sso-ProductName}:
+.. Log in to link:{user-sso-url}/auth/admin/{user-sanitized-username}/console[{sso-ProductName} administration console, window="_blank"].
+.. Click *Clients* from the side navigation.
+
+[type=verification]
+Do you see the client with the same ID that the application in 3scale?
+
+[type=verificationFail]
+Verify that you followed each step in the procedure above. If you are still having issues, contact your administrator.
+
+=== Update the client in Red Hat Single Sign-On
+
+For testing the API from the Developer Portal, we'll need to enable CORS on the {sso-ProductName} client. Normally, this will not be needed, as the OAuth handshake and token exchange will be done on the server side.
+
+. Log in to link:{user-sso-url}/auth/admin/{user-sanitized-username}/console[{sso-ProductName} administration console, window="_blank"].
+. Click *Clients* from the side navigation.
+. Select the client that was created by 3scale from the clients list by clicking on its *Client ID*.
+. At the bottom of the page, enter `+` in the *Web Origins* field. This will allow CORS for the origins of the *Valid Redirect URIs* (will be configured later). You can also enter `*` to permit all origins.
```

Comment:
Might be good to add a note explaining `+` vs `*`. It's my first time seeing it and this is what I could find[1]

Is it possible to simply enter the correct hostnames instead of `+` or `*` since that's a better practice for security reasons?

[1] - https://webcache.googleusercontent.com/search?q=cache:R_o3RSemmIQJ:https://lists.jboss.org/pipermail/keycloak-user/2017-October/012008.html+&cd=1&hl=en&ct=clnk&gl=us

Reply:
Well, I think that `+` is actually a good option, because it will only add the host of the URL that is entered in `Valid Redirect URIs`. So, in this case from the security point of view it would be as good as type in the dev portal URL in `Web Origins`.

But it is actually better, because with the explicit origins if (when) the Redirect URI changes (for example, for some real web app), the `Web Origins` will need to be updated too. But the developers (API consumers), who are the ones that will change the Redirect URI do not have access to the RH-SSO clients and update this field.

Well, to be honest, nobody should be checking or updating these clients in RH-SSO manually... This should all be transparent. It might actually be a good idea to add `webOrigins: "+"` to the [client object](https://github.com/3scale/zync/blob/bf9b1da/app/adapters/keycloak_adapter.rb#L49-L59) that Zync creates here.

It's a good point though that `*` is not a good idea. I will add a note.

Reply:
Oh I understand it better now, thank you! All good so.